src/compiler/utilities.ts

```diff
@@ -10270,3 +10270,8 @@ export function getTextOfJsxNamespacedName(node: JsxNamespacedName) {
 export function intrinsicTagNameToString(node: Identifier | JsxNamespacedName) {
     return isIdentifier(node) ? idText(node) : getTextOfJsxNamespacedName(node);
 }
+
+/** @internal */
+export function getOutermostDecoratedClass(node: Node) {
+    return findAncestor(node.parent, n => isDecorator(n) && isClassLike(n.parent) && getContainingClass(n.parent.parent) === undefined);

```

Comment:
If you're concerned about walking to the top of the tree, you can always change this to short-circuit on the first class:

```suggestion
    return findAncestor(node.parent, n => isClassLike(n) ? "stop"
        : isDecorator(n) && isClassLike(n.parent) && getContainingClass(n.parent.parent) === undefined);
```